dali/test/dali_operator_test.h

```diff
@@ -23,30 +23,197 @@
+std::vector<TensorListWrapper>
+GetOutputsFromPipeline(Pipeline &pipeline, const std::string &output_backend) {
+  std::vector<TensorListWrapper> ret;
+  auto workspace = CreateWorkspace();
+  pipeline.Outputs(&workspace);
+  for (int output_idx = 0; output_idx < workspace.NumOutput(); output_idx++) {
+    ret.emplace_back(&workspace.template Output<OutputBackend>(output_idx));
+  }
+  return ret;
+}
+
+
+void BuildPipeline(Pipeline &pipeline, const OpSpec &spec) {
+  std::vector<std::pair<std::string, std::string>> vecoutputs_;
+  for (int i = 0; i < spec.NumOutput(); ++i) {
+    vecoutputs_.emplace_back(spec.OutputName(i), spec.OutputDevice(i));
+  }
+  pipeline.Build(vecoutputs_);
+}
+
+
+// TODO(mszolucha): graph of operators
+OpSpec CreateOpSpec(const std::string &operator_name, Arguments operator_arguments, bool has_input,
+                    const std::string &input_backend, const std::string &output_backend) {
+  assert(input_backend == "cpu" || input_backend == "gpu");
+  assert(output_backend == "cpu" || output_backend == "gpu");
+
+  OpSpec opspec = OpSpec(operator_name);
+  for (auto &arg : operator_arguments) {
+    arg.second.SetArg(arg.first.arg_name(), opspec, nullptr);
+  }
+  if (has_input) {
+    opspec.AddInput("input", input_backend);
+  }
+  opspec.AddOutput("output", output_backend);
+  return opspec;
+}
+
+}  // namespace detail
+
+/**
+ * Defines and conducts test for given graph of operators.
+ * Note: whenever "graph of operators" is used, in particular it can mean,
+ *       that the graph consists of one operator.
+ *
+ * In order to set up a test, please define custom subclass of this class.
+ *
+ * Along with custom class, a verification function should be defined (Verify, VerifySiso)
+ *
+ * GenerateOperatorGraph is a template method, where you can define,
+ * what precisely should be tested.
+ *
+ * Lastly, to actually run a test, please call RunTest within GTest's testing macro
+ */
 class DaliOperatorTest : public ::testing::Test, public ::testing::WithParamInterface<Arguments> {
  public:
-  using Verify = std::function<void(TensorListWrapper, TensorListWrapper, Arguments)>;
+  DaliOperatorTest(size_t batch_size, size_t num_thread) :
+          batch_size_(batch_size), num_threads_(num_thread) {}
+
+
+  /**
+   * Type of the function used for verification, whether graph test passed or failed.
+   * Within this function, please use GTest's ASSERTs and EXPECTs, for test validation.
+   *
+   * First argument of function contains all inputs, that have been provided;
+   * Second argument are all outputs from pipeline;
+   * Third argument contains Arguments, with which the pipeline was called
+   *
+   * This is some sort or template method pattern.
+   */
+  using Verify = std::function<void(const std::vector<TensorListWrapper> /* inputs */,
+                                    const std::vector<TensorListWrapper> /* outputs */,
+                                    const Arguments)>;
```

Comment:
1. Why is `Verify` a nested type? Does it really depend that much on `DaliOperatorTest`?
2. Instead of adding another flavor of `Verify`, consider:
```
inline Verify verify_single_io(std::function<void(const TensorListWrapper /* single input */ &,
                                        const TensorListWrapper & /* single output */,
                                        const Arguments&)> verify_single_io {
  return [=](const std::vector<TensorListWrapper> &in, const std::vector<TensorListWrapper> &out, const Arguments &args) {
    DALI_ENFORCE(in.size() == 1, "This wrapper can only be in pipelines with exactly one input");
    DALI_ENFORCE(out.size() == 1, "This wrapper can only be in pipelines with exactly one output");
    return verify_single_io(in[0], out[0], args);
  };
}
```

Reply:
We decided, to keep second flavor of Verify (`VerifySingleIo`), in order to keep API simpler